streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobGenerator.scala

```diff
@@ -93,27 +93,18 @@ class JobGenerator(jobScheduler: JobScheduler) extends Logging {
 
     if (processReceivedData) {
       logInfo("Stopping JobGenerator gracefully")
-      val timeWhenStopStarted = System.currentTimeMillis()
       val stopTimeout = conf.getLong(
         "spark.streaming.gracefulStopTimeout",
         10 * ssc.graph.batchDuration.milliseconds
       )
-      val pollTime = 100
-
-      // To prevent graceful stop to get stuck permanently
-      def hasTimedOut = {
-        val timedOut = System.currentTimeMillis() - timeWhenStopStarted > stopTimeout
-        if (timedOut) {
-          logWarning("Timed out while stopping the job generator (timeout = " + stopTimeout + ")")
-        }
-        timedOut
-      }
+
 
       // Wait until all the received blocks in the network input tracker has
       // been consumed by network input DStreams, and jobs have been generated with them
       logInfo("Waiting for all received blocks to be consumed for job generation")
-      while(!hasTimedOut && jobScheduler.receiverTracker.hasUnallocatedBlocks) {
-        Thread.sleep(pollTime)
+      if (!TimeoutUtils.waitUntilDone(stopTimeout,
```

Comment:
Actually, instead of this utility, it might be a better idea to use the `clock` in this class, which has the `waitForTime` method. In fact using that clock allows us to check stuff in unit tests using manual clock. Might help you in fact in implementing a unit test for this.